proposals/proposalTypes/Proposal.sol

```diff
@@ -0,0 +1,94 @@
+//SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity 0.8.18;
+
+import {Test} from "@forge-std/Test.sol";
+import {IProposal} from "@proposals/proposalTypes/IProposal.sol";
+
+import {Core} from "@protocol/core/Core.sol";
+import {Roles} from "@protocol/core/Roles.sol";
+import {Addresses} from "@proposals/Addresses.sol";
+
+abstract contract Proposal is IProposal, Test {
+    bool public DEBUG = true;
+
+    Core _core;
+
+    function setDebug(bool value) external {
+        DEBUG = value;
+    }
+
+    /// @notice run the deployment for testing only. ie intergration testing with foundry
+    /// @dev this is not run on-chain, and is only used with the foundry `forge test` command
+    function deployForTestingOnly(Addresses addresses, address deployer) public {
+        vm.startBroadcast(deployer);
+
+        _beforeDeploy(addresses, deployer);
+        _deploy(addresses, deployer);
+        _afterDeploy(addresses, deployer);
+        _build(addresses, deployer);
+        _run(addresses, deployer);
+        _teardown(addresses, deployer);
+        _validate(addresses, deployer);
+        _validateForTestingOnly(addresses, deployer);
+
+        vm.stopBroadcast();
+    }
+
+    /// @notice run the deployment on-chain
+    /// @dev this is run on-chain and is used with the foundry `forge script` command
+    function deployOnChain(Addresses addresses, uint256 privateKey) public {
+        vm.startBroadcast(privateKey);
+        address deployer = vm.addr(privateKey);
+
+        _beforeDeploy(addresses, deployer);
+        _deploy(addresses, deployer);
+        _afterDeploy(addresses, deployer);
+        _afterDeployOnChain(addresses, deployer); // revoke admin role from deployer
+        _build(addresses, deployer);
+        _run(addresses, deployer);
```

Comment:
you probably shouldn't broadcast here either, though I'm less sure on this one. definitely do not broadcast build

run can be broadcast, but only if you are both deploying and submitting a proposal to a DAO at the same time.

usually though, you would separate these things out.

Reply:
yeah i wasnt sure about this stuff either and just through that once we needed these features we can update this code to be fit for purpose then as its hard to test stuff we dont actually use at this point.